### ISSUE TITLE
[Spaceship] cannot update IAP subscription which is already Approved

### ISSUE THREAD

**Issue by j3ski-passion** [fastlane/fastlane#15437]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
Hello, I'm trying to update an existing IAP (recurring type) and it works well if the subscription is still `Ready to Submit` with this snippet:
```ruby
      edit = iap.edit
      edit.pricing_intervals = [{
        country: "WW",
        begin_date: Date.today.to_s,
        end_date: nil,
        tier: options[:tier]
      }]
      edit.subscription_price_target = {currency: "USD", tier: options[:tier]}

      edit.save!
```

however it fails with an `ITC.response.error.OPERATION_FAILED` error even if I just try
```ruby
      edit = iap.edit
      edit.save!
```

##### Command executed
```
fastlane upsert_subscription app_id:xxx product_id:xxx tier:xxx
```

##### Complete output when running fastlane, including the stack trace and command used

(the missing plugin is not used in this lane)

<details>
  <pre> [[13:27:56]: Error loading plugin 'fastlane-plugin-appicon': Unable to activate fastlane-plugin-appicon-0.14.0, because mini_magick-4.9.5 conflicts with mini_magick (~> 4.5.1)
+------------------------+-----------+---------------------+
|                       Used plugins                       |
+------------------------+-----------+---------------------+
| Plugin                 | Version   | Action              |
+------------------------+-----------+---------------------+
| fastlane-plugin-find_  | 0.1.0     | find_replace_string |
| replace_string         |           |                     |
| fastlane-plugin-appic  | undefined | No actions found    |
| on                     |           |                     |
| fastlane-plugin-downl  | 1.0.0     | download_file       |
| oad_file               |           |                     |
| fastlane-plugin-load_  | 0.0.1     | load_json           |
| json                   |           |                     |
+------------------------+-----------+---------------------+

[13:27:57]: ------------------------------
[13:27:57]: --- Step: default_platform ---
[13:27:57]: ------------------------------
[13:27:57]: Driving the lane 'ios upsert_subscription' 🚀
[13:27:57]: -----------------------------
[13:27:57]: --- Step: setup_circle_ci ---
[13:27:57]: -----------------------------
[13:27:57]: Not running on CI, skipping CI setup
[13:28:00]: IAP found, updating
+------------------+-------------------------+
|                Lane Context                |
+------------------+-------------------------+
| DEFAULT_PLATFORM | ios                     |
| PLATFORM_NAME    | ios                     |
| LANE_NAME        | ios upsert_subscription |
+------------------+-------------------------+
[13:28:09]: ITC.response.error.OPERATION_FAILED

+------+------------------+-------------+
|           fastlane summary            |
+------+------------------+-------------+
| Step | Action           | Time (in s) |
+------+------------------+-------------+
| 1    | default_platform | 0           |
| 2    | setup_circle_ci  | 0           |
+------+------------------+-------------+

[13:28:09]: fastlane finished with errors

Looking for related GitHub issues on fastlane/fastlane...

➡️  fastlane init error
    https://github.com/fastlane/fastlane/issues/9241 [closed] 25 💬
    29 Aug 2017

➡️  Can't update in-app purchase with spanish language
    https://github.com/fastlane/fastlane/issues/9115 [closed] 11 💬
    25 Sep 2017

➡️  Binary upload invalid / [!] {"data"=>nil, "messages"=>{"warn"=>nil, "error"=>["ITC.response.error.OPERATION_FAILED"], "info"=>nil}, "statusCode"=>"ERROR"}
    https://github.com/fastlane/fastlane/issues/6419 [closed] 84 💬
    26 Apr 2017

and 4 more at: https://github.com/fastlane/fastlane/search?q=The%20request%20could%20not%20be%20completed%20because%3A%0A%09ITC.response.error.OPERATION_FAILED&type=Issues&utf8=✓

🔗  You can ⌘ + double-click on links to open them directly in your browser.

[!] The request could not be completed because:
	ITC.response.error.OPERATION_FAILED] </pre>
</details>

### Environment

<details>
  <pre> [<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 10.14.6                                     |
| Ruby                        | 2.2.4                                       |
| Bundler?                    | false                                       |
| Git                         | git version 2.21.0 (Apple Git-122)          |
| Installation Source         | ~/.fastlane/bin/bundle/bin/fastlane         |
| Host                        | Mac OS X 10.14.6 (18G103)                   |
| Ruby Lib Dir                | ~/.fastlane/bin/bundle/lib                  |
| OpenSSL Version             | OpenSSL 1.0.2g  1 Mar 2016                  |
| Is contained                | false                                       |
| Is homebrew                 | true                                        |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 11.0                                        |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE | en_US.UTF-8 | ✅ |

### fastlane files:

**No Fastfile found**

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.133.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| slack-notifier            | 2.3.2        |
| atomos                    | 0.1.3        |
| CFPropertyList            | 2.3.5        |
| claide                    | 1.0.2        |
| colored2                  | 3.1.2        |
| nanaimo                   | 0.2.6        |
| xcodeproj                 | 1.8.1        |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| terminal-notifier         | 2.0.0        |
| unicode-display_width     | 1.1.3        |
| terminal-table            | 1.7.3        |
| plist                     | 3.2.0        |
| public_suffix             | 2.0.5        |
| addressable               | 2.5.1        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| tty-screen                | 0.6.5        |
| tty-cursor                | 0.6.1        |
| tty-spinner               | 0.9.0        |
| babosa                    | 1.0.2        |
| colored                   | 1.2          |
| highline                  | 1.7.8        |
| commander-fastlane        | 4.4.6        |
| excon                     | 0.55.0       |
| faraday                   | 0.12.1       |
| faraday_middleware        | 0.11.0.1     |
| unf_ext                   | 0.0.7.4      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20170404 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| fastimage                 | 2.1.0        |
| gh_inspector              | 1.1.3        |
| json                      | 1.8.1        |
| mini_magick               | 4.9.5        |
| multi_xml                 | 0.6.0        |
| rubyzip                   | 1.3.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 0.0.4        |
| dotenv                    | 2.2.0        |
| bundler                   | 1.14.6       |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.5        |
| jwt                       | 2.1.0        |
| uber                      | 0.0.15       |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 2.1.0        |
| mime-types-data           | 3.2016.0521  |
| mime-types                | 3.1          |
| multi_json                | 1.12.1       |
| signet                    | 0.11.0       |
| memoist                   | 0.16.0       |
| os                        | 0.9.6        |
| googleauth                | 0.6.7        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.23.9       |
| google-cloud-env          | 1.0.5        |
| google-cloud-core         | 1.3.0        |
| digest-crc                | 0.4.1        |
| google-cloud-storage      | 1.16.0       |
| emoji_regex               | 1.0.1        |
| libxml-ruby               | 3.0.0        |
| psych                     | 2.0.8        |
</details>

*generated on:* **2019-10-04**] </pre>
</details>

### EDIT
I somehow made it work. I'm no rubyist, so I won't attempt to create any PRs, but this is the code that made it work.

```ruby
        client = iap.client

        intervals = client.transform_to_raw_pricing_intervals(
          app.apple_id,
          iap.purchase_id,
          nil,
          { currency: "USD", tier: options[:tier] }
        )

        intervals = intervals.map do |interval|
          interval["value"]["priceTierEffectiveDate"] = (Date.today + 1).to_s
          interval
        end

        client.update_recurring_iap_pricing!(
          app_id: app.apple_id,
          purchase_id: iap.purchase_id,
          pricing_intervals: intervals
        )
```
Key findings:
`transform_to_raw_pricing_intervals` disregards the dates passed in the intervals, so I pass in nil and set it up myself (the priceTierEffectiveDate is important when planing a price change)
iap_detail.save! breaks on [this line](https://github.com/fastlane/fastlane/blob/09ae261d854f07a233cf069f463843885f56be75/spaceship/lib/spaceship/tunes/iap_detail.rb#L199), you probably cant' do edits to a published product(?), so there should be a separate method for updating just the pricing.

Hope this helps!

**Comment by janpio** [comment]
Ping @max-ott

**Comment by frankiesimon** [comment]
This happens to me too. Code that used to update IAPs doesn't work anymore.

**Comment by max-ott** [comment]
@j3ski-passion @frankiesimon Does this only happen for `live subscriptions` or also other IAPs? I currently do not have a live subscription to test this.

**Comment by j3ski-passion** [comment]
@max-ott This only happens for `live subscriptions`, for `non-live` I am able to update them with the `iap_detail.save!` method.

**Comment by max-ott** [comment]
I'm finally able to reproduce this issue, but will need to look into it in more detail. I suspect that `spaceship` is sending the response in a different format than App Store Connect expects it or simply misses out a new parameter. I'll try to inspect the data with Charles when changing it on ASC and then compare it.

**Comment by max-ott** [comment]
Okay, so there is a main difference between editing an `approved/live` recurring subscription and an `unapproved` one. For the latter you can simply submit the complete data via `"ra/apps/#{app_id}/iaps/#{purchase_id}/pricing/subscriptions"` for all countries. 

Once you have an approved version, this no longer works. If you change the price, App Store Connect submits a `PUT` request for each country, containing both the old and the new price in an array, basically: `"ra/apps/#{app_id}/iaps/#{purchase_id}/pricing/subscriptions/US` 

```
pricing_data["subscriptions"] = [{
              ### This is basically the existing/old price
              "value" => {
                "tierStem" => "131",
                "priceTierEffectiveDate" => nil,
                "priceTierEndDate" => "2019-11-09",
                "country" => "AL",
                "grandfathered" => {
                  "value" => "FUTURE_GF",
                  "isEditable" => true,
                  "isRequired" => false,
                  "errorKeys" => nil
                }
              },
              "isEditable" => false,
              "isRequired" => false,
              "errorKeys" => nil
            }, {
              ### This will become the new/planned price
              "value" => {
                "tierStem" => "100",
                "priceTierEffectiveDate" => "2019-11-10",
                "priceTierEndDate" => nil,
                "country" => "AL",
                "grandfathered" => {
                  "value" => "FUTURE_NONE",
                  "isEditable" => false,
                  "isRequired" => false,
                  "errorKeys" => nil
                }
              },
              "isEditable" => true,
              "isRequired"=> false,
              "errorKeys" => nil
            }]
```

So basically https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/tunes/tunes_client.rb#L1319 needs to get split in 2 methods: `create` and `update` (Naming of the current method is not great as it's used on the initial setup as well). Then we would need to track down a few other places where those methods are used and use the correct one. This is quite an amount of work, so again I hope with these insights a more proficient `ruby` expert can do the magic.

**Comment by janpio** [comment]
nah
